### PR TITLE
aws profile, describe stack, custom json updates

### PR DESCRIPTION
**Pull request by ahmedn-quandl**
- Add in support to deploy using stack and layer names.
- pull relevant information from AWS stack settings ( chef version/ os version )
- Add in support to be able to switch AWS profile
- Add command to display opsworks necessary info, which fetched from AWS

### CODE REVIEW

File: lib/qops/cookbook/cookbook.rb [hunk 1]
```diff
@@ -130,10 +130,19 @@ def config
...
-      region: 'us-east-1',
-      credentials: config.opsworks.config.credentials.credentials
-    )
+    @s3 ||= Aws::S3::Client.new(**aws_configs)
+  end
+
+  def aws_configs
+    aws_config = { region: 'us-east-1' }
+    # use the profile if found
+    if options[:profile]
```
Comment (junos) [inline lib/qops/cookbook/cookbook.rb]: you can use one-liner like this to make this simple: aws_config = options[:profile] ? A : B

File: lib/qops/deployment/app.rb [hunk 2]
```diff
@@ -103,15 +103,17 @@ def custom_json
...
-      @_custom_json[:deploy] = {
-        revision: revision_used,
-        # the following is to be deprecated in favor of the former...
-        application_name => {
-          scm: {
-            revision: revision_used
-          }
-        }
-      }
+      @_custom_json[:deploy] = if config.chef_version >= 12.2
```
Comment (junos) [inline lib/qops/deployment/app.rb]: why this matter for chef version?

File: lib/qops/environment.rb [hunk 3]
```diff
@@ -31,20 +31,75 @@ def self.print_with_colour(message, level = :normal)
...
+    end
+
+    def find_stack(options = {})
+      return @_stack if @_stack
+      key = if !options[:name].nil?
+              :name
+            elsif !options[:stack_id].nil?
+              :stack_id
+            else
+              identity_from_config
```
Comment (junos) [inline lib/qops/environment.rb]: could you put print statement to notify user they are using the opswork config here?

File: lib/qops/environment.rb [hunk 4]
```diff
@@ -31,20 +31,75 @@ def self.print_with_colour(message, level = :normal)
...
+      stack = opsworks.describe_stacks.stacks.find { |s| s.send(key) == value }
+      unless stack
+        puts Rainbow("Could not find stack with #{key} = #{value}").bg(:black).red
+        exit(-1)
+      end
+      @_stack = stack
+    end
+
+    def stack_id(options = {})
+      configuration.stack_id || find_stack(options).stack_id
```
Comment (junos) [inline lib/qops/environment.rb]: I think you should update this and following statement to use either config or find by stack.

File: lib/qops/environment.rb [hunk 5]
```diff
@@ -31,20 +31,83 @@ def self.print_with_colour(message, level = :normal)
...
+      find_stack(options).stack_id
+    end
+
+    def subnet(options = {})
+      return configuration.subnet if @_force_config
+      find_stack(options).default_subnet_id
+    end
+
+    def layers(options = {})
+      opsworks.describe_layers(stack_id: find_stack(options).stack_id).layers
```
Comment (junos) [inline lib/qops/environment.rb]: why layers has not force_config check?

Comment (ahmedn-quandl) [reply lib/qops/environment.rb]: layers is not used in the opsworks.yml, only layer_id. this is really more of a helper method used when force_config is set to false, and qops will search for the layers with the same stack id. Layer id enforces force_config check .

File: lib/qops/environment.rb [hunk 6]
```diff
@@ -31,20 +31,83 @@ def self.print_with_colour(message, level = :normal)
...
+            end
+      value = options[key] || configuration.send(key)
+      key = :name if key == :stack_name
+      stack = opsworks.describe_stacks.stacks.find { |s| s.send(key) == value }
+      unless stack
+        puts Rainbow("Could not find stack with #{key} = #{value}").bg(:black).red
+        exit(-1)
+      end
+      @_stack = stack
+    end
```
Comment (junos) [inline lib/qops/environment.rb]: this function is a little big and hard to understand, do you mind to break it into several short and easy to understand functions?

**Pull request merged**